In [2]:
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import numpy as np

<h1><b>B. Khám phá dữ liệu (đan xen với tiền xử lý dữ liệu)</b></h1>

### Đọc dữ liệu từ tập tin (file)

Đầu tiên, đọc dữ liệu từ file "gold.csv" và lưu kết quả vào DataFrame df_gold.

In [3]:
df_gold = pd.read_csv('gold.csv')
df_gold.head()

Date   Open   High    Low  Close  Volume Currency
0  2000-01-04  289.5  289.5  280.0  283.7   21621      USD
1  2000-01-05  283.7  285.0  281.0  282.1   25448      USD
2  2000-01-06  281.6  282.8  280.2  282.4   19055      USD
3  2000-01-07  282.5  284.5  282.0  282.9   11266      USD
4  2000-01-10  282.4  283.9  281.8  282.7   30603      USD

### Dữ liệu có bao nhiêu dòng và bao nhiêu cột

In [4]:
df_gold.shape

(5703, 7)

Dữ liệu có 5703 dòng và 7 cột

### Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

Mỗi dòng trong DataFrame `df_vege` cho biết thông tin của giá vàng trong một này, bao gồm ngày, khối lượng bán ra,.. Có vẻ không có vấn đề các dòng có ý nghĩa khác nhau.

### Dữ liệu có các dòng bị lặp không?

Ta sẽ xem thử rằng trong tập có dòng nào bị trùng không. Nếu trùng thì trả về True, không thì là False và lưu số dòng trùng vào biến num_duplicated_rows

In [5]:
#Tính tổng dòng lặp
num_duplicated_rows = df_gold.duplicated().sum()
is_duplicated = (num_duplicated_rows != 0)
print('So dong lap la {}'.format(num_duplicated_rows))
if is_duplicated == False:
    print('Bo du lieu khong lap')

So dong lap la 0
Bo du lieu khong lap


Vậy không có dòng lặp trong bộ dữ liệu này

### Mỗi cột có ý nghĩa gì?

Ý nghĩa của mỗi cột là:

- **Date** là ngày giao dịch của giá vàng đó

- **Open**  là giá mở cửa của vàng vào ngày giao dịch.

- **High** là giá cao nhất của vàng trong ngày giao dịch đó.

- **Low** là giá thấp nhất của vàng trong ngày giao dịch đó.

- **Close**  là giá đóng cửa của vàng vào ngày giao dịch đó.

- **Volume**  là số lượng vàng được giao dịch trong ngày đó.

- **Currency** là đơn vị tiền tệ được sử dụng để định giá vàng.


In [6]:
df_gold.columns

Index(['Date', 'Open', 'High', 'Low', 'Close', 'Volume', 'Currency'], dtype='object')

Ta kiểm tra dạng dữ liệu của các cột để xem có đúng với dạng mà ta mong muốn không.

In [7]:
df_gold.dtypes

Date         object
Open        float64
High        float64
Low         float64
Close       float64
Volume        int64
Currency     object
dtype: object

Ta chuyển cột Date sang dạng datetime

In [8]:
df_gold['Date'] = pd.to_datetime(df_gold['Date'],format='%Y-%m-%d')

## XỬ LÝ DỮ LIỆU NAN

Ta xem thử mỗi cột có bao nhiêu dữ liệu nan

In [9]:
df_gold.isnull().sum()

Date        0
Open        0
High        0
Low         0
Close       0
Volume      0
Currency    0
dtype: int64

## Với mỗi cột có kiểu dữ liệu dạng số, các giá trị được phân bố như thế nào?

In [10]:
dtypes = df_gold.dtypes
ind = dtypes[(dtypes == np.float64) | (dtypes == np.int64)].index
ind

Index(['Open', 'High', 'Low', 'Close', 'Volume'], dtype='object')

Ta có các cột dữ liệu dạng số 'Rating', 'Year', 'Runtime', 'Budget', 'Income'. Ta sẽ tính min, lower quartile, median, upper quartile, max để xem sự phân bố của dữ liệu và có gì bất thường ở các cột này không?

In [11]:
def five_summary(df_gold):
    num_col_info_df = pd.DataFrame([],index= ["min", "lower_quartile", "median", "upper_quartile", "max" ])
    for i in ind:
        data = np.nanpercentile(df_gold[i],(0,25,50,75,100)) #tinh min, lower quartile, median, upper quartile, max
        num_col_info_df[i] = data.round(1)
    return num_col_info_df
five_summary(df_gold)

Open    High     Low   Close    Volume
min              256.6   259.4   255.1   256.6       0.0
lower_quartile   459.8   463.9   457.4   460.5   52938.5
median          1188.8  1198.0  1179.7  1188.7  126006.0
upper_quartile  1381.4  1392.8  1368.1  1383.0  193109.0
max             2076.4  2089.2  2049.0  2069.4  816531.0

Ta thấy giá trị min của Volume bằng 0, đây có thể là một lỗi, vì vậy ta nên xem thử

In [12]:
df_gold[df_gold['Volume'] == 0]

Date    Open     High      Low   Close  Volume Currency
223  2000-11-21   267.0   267.60   265.90   266.1       0      USD
276  2001-02-09   262.2   263.00   260.70   262.0       0      USD
277  2001-02-12   262.7   263.40   261.90   262.7       0      USD
300  2001-03-16   260.0   261.20   258.00   258.4       0      USD
322  2001-04-18   262.0   263.20   259.20   262.0       0      USD
422  2001-09-10   275.0   275.80   272.70   273.7       0      USD
602  2002-06-04   327.7   331.50   327.70   328.8       0      USD
622  2002-07-02   314.3   316.00   313.00   313.2       0      USD
724  2002-11-26   318.1   318.90   317.10   317.7       0      USD
5702 2022-09-02  1707.9  1729.45  1707.05  1723.0       0      USD

Volume = 0 có thể là do phiên giao dịch không mở cửa hay do lỗi của quá trình thu thập dữ liệu, vì số lượng không quá nhiều nên ta có thể xoá nó đi để thuận tiện cho việc phân tích

In [13]:
df_gold.drop(df_gold[df_gold['Volume'] == 0].index,inplace = True)

In [ ]:
df_gold.to_csv('gold_clean.csv')